packaging/linux/scripts/after-install-and-upgrade.sh

```diff
@@ -97,9 +97,19 @@ fi
 
 # Restart on upgrade, if we can
 if [ "$OPERATION" = "upgrade" ] ; then
-  if [ $BK_SYSTEMD_EXISTS -eq 0 ] && systemctl is-active buildkite-agent --quiet; then
-    # Try using systemd, if the unit is active
-    systemctl try-restart buildkite-agent
+  # Try restarting with systemd, if the unit is active
+  if [ $BK_SYSTEMD_EXISTS -eq 0 ]; then
+    if systemctl is-active buildkite-agent --quiet; then
+      systemctl try-restart buildkite-agent
+    fi
+
+    # For instantiations of the template. Trying to find one pattern that covers
+    # all instantiations and the non-templated version is possible, it is
+    # `buildkite-agent*`, however that would also cover other service names such
+    # as `buildkite-agentless`. It's safer to do two, more restrictive restarts.
+    if systemctl is-active 'buildkite-agent@*' --quiet; then
+      systemctl try-restart 'buildkite-agent@*'
+    fi
```

Comment:
I think this has a secondary change that means folks with systemd installed but not using it to manage buildkite-agent might not see a restart after this change.

There's also no need to check is-active when using try-restart, so if you want to double down on this conditional you can just do a try-restart command.

But I'd suggest:

```suggestion
  # Try restarting with systemd, if the unit is active, either as a base unit or as an instance of the template.
  if [ $BK_SYSTEMD_EXISTS -eq 0 ] && ( systemctl is-active buildkite-agent 'buildkite-agent@*' --quiet ); then
    systemctl try-restart buildkite-agent 'buildkite-agent@*'
```

Reply:
> I think this has a secondary change that means folks with systemd installed but not using it to manage buildkite-agent might not see a restart after this change where they would have via the pkill previously.

Good catch. I'll fix.

For the suggestion, on my machine, the pattern that matches the instances of the template unit does not match the base unit and vice versa. So I think 2 of everything is needed.

We could have an `||` with `is-active`.

Reply:
Yeah, neat. Using some `||` magic for the detection would be good. I don't think there's any harm to running two `try-restart` commands without switching between, they're meant to fail silently if the unit isn't running.

Reply:
I avoided the `||` because I don't think I know how to make parens work in bash, and it would be too long without factoring out the systemd exsistence check.

Reply:
Hmm, some people might have both the base unit and the template instances started, and this won't catch the latter in that case. I'll think of something else then.

Reply:
Luckily, I think putting the patterns in the same `systemctl` invocation will work.